django/db/models/aggregates.py

```diff
@@ -85,6 +87,10 @@ def as_sql(self, compiler, connection, **extra_context):
                 condition = When(self.filter, then=source_expressions[0])
                 copy.set_source_expressions([Case(condition)] + source_expressions[1:])
                 return super(Aggregate, copy).as_sql(compiler, connection, **extra_context)
+        if self.default is not None:
+            clone = self.copy()
+            clone.default = None  # Reset the default argument before wrapping.
+            return Coalesce(clone, self.default).as_sql(compiler, connection, **extra_context)
```

Comment:
Any thoughts about doing that at `resolve_expression` time instead where a `copy` is already created?

Looks like the same logic could be applied there and avoid unnecessary copies on `as_sql`

e.g.

```python
def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
    ...
    if c.default is not None:
        default = c.default
        c.default = None
        return Coalesce(c, default)
    return c
```

The reason why `filter` handling as to be done in `as_sql` is because it fundamentally changes the generated SQL based on whether or not `connection.features.supports_aggregate_filter_clause` but it's not the case for coalescence.

Reply:
Ah, thanks. Sounds like a good plan - still learning bits of the expressions API.

Unfortunately moving to this approach this breaks things for `Avg`, `StdDev`, and `Variance` which all use `NumericOutputFieldMixin` to force returning `DecimalField` or `FloatField`:

```
django.core.exceptions.FieldError: Expression contains mixed types: FloatField, IntegerField. You must set output_field.
```

So it looks like output field resolution doesn't get triggered in the same way...

> Having a test to make sure things work appropriately with `filter=` would also be valuable I believe.

Good idea. Will add one.

Reply:
@pope1ni that's expected, you'll want to pass `output_field=c._output_field_or_none` as well so

```python
def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
    ...
    if c.default is not None:
        default = c.default
        c.default = None
        return Coalesce(c, default, output_field=c._output_field_or_none)
    return c
```

Reply:
Ah, gotcha.